hot_ui/lib/main.dart

```diff
@@ -0,0 +1,244 @@
+  final String isolateId;
+
+  Future<void> reloadMethod(String libraryId, String classId, String methodId,
+      String methodBody) async {
+    final Stopwatch sw = Stopwatch()..start();
+    final Procedure procedure = await incrementalCompiler
+      .compileExpression(methodBody, <String>['context'], <String>['BuildContext'], libraryId, classId, false);
+    component.transformChildren(BodyReplacementTransformer(libraryId, classId, methodId, procedure));
+    final Component partialComponent = Component(libraries: <Library>[
+      component.libraries.firstWhere((Library library) => library.importUri.toString() == libraryId)
+    ]);
+    stderr.writeln('compiled expression in ${sw.elapsedMilliseconds}');
+
+    // Use the HTTP request as the sink for serializing the partial dill.
+    final HttpClientRequest request =
+        await httpClient.putUrl(Uri.parse(httpAddress));
+    request.headers.removeAll(HttpHeaders.acceptEncodingHeader);
+    request.headers.add('dev_fs_name', devFs);
+    request.headers.add(
+        'dev_fs_uri_base64',
+        base64
+            .encode(utf8.encode('${devfsUri}lib/main.dart.incremental.dill')));
+    final BinaryPrinter binaryPrinter = BinaryPrinter(request);
+    binaryPrinter.writeComponentFile(partialComponent);
+    await request.close();
+    await vmService.reloadSources(isolateId);
+
+    // Invoke reassemble.
+    await vmService.callServiceExtension('flutter.ext.reassemble');
+  }
+
+  // Expects newline denominated JSON object.
+  void onMessage(String line) {
+    final Map<String, Object> message = json.decode(line);
+    final String libraryId = message['libraryId'];
+    final String classId = message['classId'];
+    final String methodId = message['methodId'];
+    final String methodBody = message['methodBody'];
+
+    reloadMethod(libraryId, classId, methodId, methodBody).then((void result) {
+      print(kReadyMessage);
+    }).catchError((dynamic error, StackTrace stackTrace) {
+      stderr.writeln(error.toString());
+      stderr.writeln(stackTrace.toString());
+      print(kFailedMessage);
+    });
+  }
+}
+
+/// Replaces the body of method [methodId] on class [classId] in library
+/// [libraryId] with [procedure].
+class BodyReplacementTransformer extends Transformer {
+  BodyReplacementTransformer(
+    this.libraryId,
+    this.classId,
+    this.methodId,
+    this.procedure,
+  );
+
+  final String libraryId;
+  final String classId;
+  final String methodId;
+  final Procedure procedure;
+
+  @override
+  Library visitLibrary(Library node) {
+    if (node.importUri.toString() == libraryId) {
+      return super.visitLibrary(node);
+    }
+    return node;
+  }
+
+  @override
+  Class visitClass(Class clazz) {
+    if (clazz.name == classId) {
+      return super.visitClass(clazz);
+    }
+    return clazz;
+  }
+
+  @override
+  TreeNode visitProcedure(Procedure node) {
+    if (node.name.name == methodId && node.kind == ProcedureKind.Method) {
+      return procedure;
```

Comment:
nice!
A plausible TODO could be to throw if the procedure has different arguments than before.